source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc

```diff
@@ -165,6 +189,46 @@ void UpstreamProxyProtocolSocketFactory::hashKey(
   }
 }
 
+void UpstreamProxyProtocolSocket::processCustomTLVsFromHost() {
+  const auto& upstream_info = callbacks_->connection().streamInfo().upstreamInfo();
+  if (upstream_info == nullptr || upstream_info->upstreamHost() == nullptr) {
+    return;
+  }
+  auto metadata = upstream_info->upstreamHost()->metadata();
+  if (metadata == nullptr) {
+    return;
+  }
+
+  const auto filter_it = metadata->typed_filter_metadata().find(
+      Envoy::Config::MetadataFilters::get().ENVOY_TRANSPORT_SOCKETS_PROXY_PROTOCOL);
+  if (filter_it == metadata->typed_filter_metadata().end()) {
+    ENVOY_LOG(trace, "No custom TLVs found in upstream host metadata");
+    return;
+  }
+
+  ProxyProtocolConfig tlvs_metadata;
+  auto status = absl::OkStatus();
+  status = MessageUtil::unpackTo(filter_it->second, tlvs_metadata);
```

Comment:
This will be relatively slow at runtime. Could you dig the fields out by name and directly populate the data into the custom_tlvs vector?

Reply:
@ggreenway I added a TODO after pushing the refactored implementation. Are there any examples on how to do this? I'll look into this more today.

Reply:
@ggreenway We're grabbing the host configuration from typed_filter_metadata (vs. filter_metadata) which doesn't have access to the metadata struct fields. I can either create a wrapper message for the TlvEntry list and unpack directly to that object, but I'm guessing the performance for that implementation is also negligible compared to the current implementation that unpacks to the top-level config.